File: docs/dp_tabulation/0918_medium/circular_subarray_sum.py

def compute_max_circular_subarray_sum(nums: list[int]) -> int:  # LeetCode Q.918.
    array_total_sum = 0
    max_num = nums[0]

    pos_subarray_sum, max_pos_sum = 0, 0

    neg_subarray_sum, min_neg_sum = 0, 0

    for num in nums:
        if num > max_num: max_num = num

        array_total_sum += num

        pos_subarray_sum += num
        if pos_subarray_sum > max_pos_sum:
            max_pos_sum = pos_subarray_sum

        if pos_subarray_sum <= 0:  # Positive sum <= 0: reset subarray.
            pos_subarray_sum = 0
        
        neg_subarray_sum += num
        if neg_subarray_sum < min_neg_sum:
            min_neg_sum = neg_subarray_sum

        if neg_subarray_sum >= 0:  # Negative sum >= 0: reset subarray.
            neg_subarray_sum = 0

    if max_num <= 0: return max_num  # All numbers are non-positive.

    return max(array_total_sum - min_neg_sum, max_pos_sum)

File: docs/dp_tabulation/0918_medium/circular_subarray_sum.cpp
#include <vector>
using namespace std;

int computeMaxCircularSubarraySum(vector<int>& nums) { // LeetCode Q.918.
    int arrayTotalSum = 0;
    int maxNum = nums.front();

    int posSubarraySum = 0, maxPosSum = 0;

    int negSubarraySum = 0, minNegSum = 0;

    for (const auto& num : nums) {
        if (num > maxNum)
            maxNum = num;

        arrayTotalSum += num;

        posSubarraySum += num;
        if (posSubarraySum > maxPosSum)
            maxPosSum = posSubarraySum;

        if (posSubarraySum <= 0) // Positive sum <= 0: reset subarray.
            posSubarraySum = 0;

        negSubarraySum += num;
        if (negSubarraySum < minNegSum)
            minNegSum = negSubarraySum;

        if (negSubarraySum >= 0) // Negative sum >= 0: reset subarray.
            negSubarraySum = 0;
    }

    if (maxPosSum == 0) // All numbers are non-positive.
        return maxNum;

    return max(arrayTotalSum - minNegSum, maxPosSum);
}
